serf/internal_query.go

```diff
@@ -149,17 +150,60 @@ func (s *serfQueries) handleConflict(q *Query) {
 	}
 }
 
-// sendKeyResponse handles responding to key-related queries.
-func (s *serfQueries) sendKeyResponse(q *Query, resp *nodeKeyResponse) {
-	buf, err := encodeMessage(messageKeyResponseType, resp)
-	if err != nil {
-		s.logger.Printf("[ERR] serf: Failed to encode key response: %v", err)
-		return
+func (s *serfQueries) keyListResponseWithCorrectSize(q *Query, resp *nodeKeyResponse) ([]byte, messageQueryResponse, error) {
+	warned := false
+	for i := len(resp.Keys); i >= 0; i-- {
+		buf, err := encodeMessage(messageKeyResponseType, resp)
+		if err != nil {
+			return nil, messageQueryResponse{}, err
+		}
+
+		// Create response
+		qresp := q.createResponse(buf)
+
+		// Encode response
+		raw, err := encodeMessage(messageQueryResponseType, qresp)
+		if err != nil {
+			return nil, messageQueryResponse{}, err
+		}
+
+		// Check the size limit
+		if err = q.checkResponseSize(raw); err != nil {
+			resp.Keys = resp.Keys[0:i]
+			if !warned {
+				s.logger.Printf("[WARN] serf: truncated key list list response so that it fits into message")
+				warned = true
+			}
+			continue
+		}
+		return raw, qresp, nil
```

Comment:
So it looks like the payload is: https://github.com/hashicorp/serf/blob/6a7f035e9efd9b99d33ee452f0b26501faec7d52/serf/internal_query.go#L51-L60

Which I notice two things:

 1. It has a Message field - could we return the fact that the truncation occured there? It would allow us to be sure in the client rather than guessing if it's close. (It adds bytes though of course.
 2. It's pretty simple although it is msgPack encoded which is not totally trivial (varints for example). In this case everything is a string or fixed length though and we could rely on all keys being the same length too so we can probably make a hand-wavy heuristic for how many keys will fit:
  - 1 byte for the struct/map header (since it's only 3 elements)
  - 1 byte for (short) string key header + 6 bytes for "result"
  - 1 byte for the bool
  - 1 byte for (short) string key header + 7 bytes for "message"
  - 1 byte for the empty string header (plus more if we use this for error per 1 above)
  - 1 byte for the (short) string key header + 4 bytes for "keys"
  - 5 bytes for the list header (3 bytes is enough up to 65k keys but be conservative)
  - so far I make that 28 bytes with empty Message
  - N * 1 byte for string header (keys are shorter than 31 bytes in base64) + 24 byte for base64 key with padding

You could add some fudge factors there but my estimate is that if the max query size is default 1024 then you can get 39 keys in.

I think in practice people were seeing issues with a few less - it's not clear if that's down to me missing something above (maybe we already use the Message for something).

I don't know if it makes sense to try to to exactly calculate message size, but my point is we know it's about 30 something keys so if there are 10,000 keys present then iterating and re-encoding 9,960+ times is not super clever!

We could maybe have a really simple model that uses big fudge factors (like `QueryResponseSizeLimit/24` where 24 is minimum number of bytes for a 16 byte key in base64) as the starting point (because we know for sure we can't encode more than that. Then we'd only have to iterate at worst a handfull of times rather than unbounded?

Reply:
Thanks for suggestion! I wrote a test that estimates that factor: TestSerfQueries_estimateMaxKeysInListKeyResponseFactor. I took an unrealistic message size of 10240 to make sure we err on the side of being to generous and also to account for other configurations.

In case of the default size, the algorithm starts at index 40, no matter how many keys there are.